### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] <Using type assertion on Object.values(Record<T> foo) gets wrongly flagged>

### ISSUE THREAD

**Issue by ilia-kebets-sonarsource** [typescript-eslint/typescript-eslint#5808]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=KYDwDg9gTgLgBAMwK4DsDGMCWEV0wZwAUJ9MsA3YAChAC44UkBbAI2CgEo4BvAKDjhRgMJFFwg4APjgAGANy8Avr16hIsODACeYYHGKkKwAHLM2UAErA00ACYAeACpxQMYClv44+GFEwoAczgAHwYzdhDvLVYIABtpAF5+OCsbKAdHABow1nZpADIeOAB9Yu1degBySENMSlNcy2s7SrhFBVVwaHgbFB88IhIyOpNwprTbOAS4JxcQNw8vHz9AyMZGyPxoljjJZKpyAENYpGB6VLsnbPXzaVDGWNjI1FtgBH9gWw56I5O9An0QyMDXMF3STkS0j4yUwCDgVAAhL9Tlw+AIBEIRGJEMd8MAFAJlAJev1kcAvNMAPIsABW1hgADoyfgDscUXBDl4buwANoAXTkAgA9EK4AAxQgDOAAC3YwBUGOEolwzIZwEoUC0VAIBmGlA4CnaQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
export function isPositive(x: number) {
  return x > 0;
}

export type PositiveNumberRecord<T extends string | number | symbol> =
  Record<T, number> & { __type: 'positiveNumberRecord' };

export const isPositiveNumberRecord = <T extends string | number | symbol>
  (value: Record<T, number> | null | undefined): value is PositiveNumberRecord<T> => {

    if (!value) {
      return false;
    }

    const values = Object.values(value) as number[];  // FP is here

    return values.every(isPositive);
};
```

### ESLint Config

```javascript
module.exports = {
  parser: '@typescript-eslint/parser',
  parserOptions: {
    tsconfigRootDir: __dirname,
    project: ['./tsconfig.json']
  },
  plugins: ['@typescript-eslint'],
  rules: {},
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-requiring-type-checking'
  ]
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
      "module": "commonjs",
      "target": "es6",
      "lib": [
        "ES2017"
      ]
  },
  "include": [
      "src/index.ts"
  ]
}
```

### Expected Result

I don't expect any error since the cast is necessary to change `unknown[]` to `number[]`.

### Actual Result

I'm getting an error:

```

  15:20  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.40.0` |
| `@typescript-eslint/parser`            | `5.40.0` |
| `TypeScript`                           | `4.8.4` |
| `ESLint`                               | `8.25.0` |
| `node`                                 | `16.16.0` |

**Comment by bradzacher** [comment]
`Object.values` has a definition that looks like this:

```ts
function values<T>(
  obj: Record<unknown, T>,
): T[];
```

When doing `Object.values(value) as number[]`, you're assuming that TS is doing is a type operation like `unknown[] as number[]`.

However what TS actually does under the hood is not this. TS recognises that the function has a generic in the return type position, so what it does is that it will attempt to sub your assertion in for the generic and see if the code still works.

Put another way - TS reworks your code as written from
```ts
Object.values(value) as number[]
// to
Object.values<number>()
```
[You can see this by inspecting the types using intellisense](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAMwK4DsDGMCWEV0wZwAUJ9MsA3YAChAC44UkBbAI2CgEo4BvAKDjhRgMJFFwg4APjgAGANy8Avr16hIsODACeYYHGKkKwAHLM2UAErA00ACYAeACpxQMYClv44+GFEwoAczgAHwYzdhDvLVYIABtpAF5+OCsbKAdHABow1nZpADIeOAB9Yu1degBySENMSlNcy2s7SrhFBVVwaHgbFB88IhIyOpNwprTbOAS4JxcQNw8vHz9AyMZGyPxoljjJZKpyAENYpGB6VLsnbPXzaVDGWNjI1FtgBH9gWw56I5O9An0QyMDXMF3STkS0j4yQEmAQcCoAEJfqcuHwBBjBMJRLgEMd8MAFJjlDC4L1+ijgF5pgB5FgAK2sMAAdJT8Adjqi4IcvDd2ABtAC6cgEAHpRXAAGKEAZwAAW7GApPFmNVarVAD0APyklSYoQiMRwNnM4CUKBaKgEAzDSgcBTtXhAA)

Why does this matter? Well our rule works entirely "under the hood". We only see the types TS tells us (I.e. We see exactly what you would see when hovering over `Object.values`). So what happens is we inspect the return type of your call and see that TS tells us that it's `number[]` - so you've tried to assert `number[]` to a `number[]`... That's unnecessary!!

Can't we work around this? No. TS has no tooling to allow us to backtrack or ask "what if" questions. So we have no way of knowing that when you delete the assertion it will change the return type to `unknown[]`.

The workaround is to use the generic directly instead (`Object.values<number>()`), or use a disable comment.
